lib/core/kotekanTrackers.cpp

```diff
@@ -0,0 +1,153 @@
+#include <errno.h>     // for errno
+#include <fstream>     // for ofstream, ostream
+#include <functional>  // for _Bind_helper<>::type, _Placeholder, bind, _1, placeholders
+#include <stdexcept>   // for runtime_error
+#include <stdlib.h>    // for exit
+#include <type_traits> // for enable_if<>::type
+#include <unistd.h>    // for gethostname
+#include <utility>     // for pair
+
+namespace kotekan {
+
+KotekanTrackers::KotekanTrackers() {}
+
+KotekanTrackers::~KotekanTrackers() {
+    restServer::instance().remove_get_callback("/trackers");
+}
+
+KotekanTrackers& KotekanTrackers::instance() {
+    static KotekanTrackers _instance;
+    return _instance;
+}
+
+void KotekanTrackers::register_with_server(restServer* rest_server) {
+    using namespace std::placeholders;
+    rest_server->register_get_callback("/trackers",
+                                       std::bind(&KotekanTrackers::trackers_callback, this, _1));
+    rest_server->register_get_callback(
+        "/trackers_current", std::bind(&KotekanTrackers::trackers_current_callback, this, _1));
+}
+
+void KotekanTrackers::trackers_callback(connectionInstance& conn) {
+    nlohmann::json return_json = {};
+
+    for (auto& stage_itr : trackers) {
+        for (auto& tracker_itr : trackers[stage_itr.first]) {
+            return_json[stage_itr.first][tracker_itr.first] = tracker_itr.second->get_json();
+        }
+    }
+
+    conn.send_json_reply(return_json);
+}
+
+void KotekanTrackers::trackers_current_callback(connectionInstance& conn) {
+    nlohmann::json return_json = {};
+
+    for (auto& stage_itr : trackers) {
+        for (auto& tracker_itr : trackers[stage_itr.first]) {
+            return_json[stage_itr.first][tracker_itr.first] =
+                tracker_itr.second->get_current_json();
+        }
+    }
+
+    conn.send_json_reply(return_json);
+}
+
+std::shared_ptr<StatTracker> KotekanTrackers::add_tracker(std::string stage_name,
+                                                          std::string tracker_name,
+                                                          std::string unit, size_t size,
+                                                          bool is_optimized) {
+    if (stage_name.empty()) {
+        ERROR_NON_OO("Empty stage name. Exiting.");
+        throw std::runtime_error("Empty stage name.");
+    }
+    if (tracker_name.empty()) {
+        ERROR_NON_OO("Empty tracker name. Exiting.");
+        throw std::runtime_error("Empty tracker name.");
+    }
+    if (unit.empty()) {
+        ERROR_NON_OO("Empty unit for tracker {:s}:{:s}. Exiting.", stage_name, tracker_name);
+        throw std::runtime_error(
+            fmt::format(fmt("Empty unit name: {:s}:{:s}"), stage_name, tracker_name));
+    }
+
+    std::shared_ptr<StatTracker> tracker_ptr =
+        std::make_shared<StatTracker>(tracker_name, unit, size, is_optimized);
+
+    std::lock_guard<std::mutex> lock(trackers_lock);
```

Comment:
I would also lock when removing and reading them as well.  Very unlikely to be a contention there, but it is theoretically possible.